### DISCUSSION TITLE
How to use manifest.py in self-hosted module repository to get .mpy files?

### DISCUSSION THREAD

**Discussion by DavesCodeMusings** [micropython/micropython#11693] [Libraries & Drivers] [answered]
I'm trying to figure out how I can use _manifest.py_ to have `mpremote mip install` take my _.py_ Python file from my [GitHub repo](https://github.com/DavesCodeMusings/thimble) and convert it to an _.mpy_ bytecode file when it's installed under /lib on the device.

For example, I recently used mpremote to do a mip install of ssd1306 display drivers. In the [micropython-lib GitHub repo](https://github.com/micropython/micropython-lib/tree/master/micropython/drivers/display/ssd1306) the only files I see are _ssd1306.py_ and _manifest.py_. Yet on my device's flash filesystem, I see _ssd1306.mpy_. Furthermore, _ssd1306.py_ (on GitHub) is plaintext readable, whereas _ssd1306.mpy_ (on my device) is not.

What's the magic ingredient that makes this happen?

I thought it was _manifest.py_. Looking at the [docs](https://docs.micropython.org/en/latest/reference/manifest.html) and the ssd1306 directory on GitHub it looks like the following should do the trick in my case:

```
metadata(description="Thimble web framework.", version="0.1.0")
module("thimble.py")
```

>I basically used the _manifest.py_ from ssd1306 and changed it to suit my needs.

But, when I use `mpremote mip install` for [my own module](https://github.com/DavesCodeMusings/thimble#how-do-i-install-it), I still get the _.py_ file with the plaintext Python code.

Is there something else I need to make this work? Something missing in my _package.json_ maybe? Or is this something that only works for the micropython-lib and not self-hosted?

Any pointers are appreciated.

**Comment by ned-pcs** [comment]
`manifest.py` is used as part of the MicroPython build process. Any Python `.py` files referred to in that manifest get compiled to `.mpy` files and built into the Micropython flash image that is later installed on your device. But those `.mpy` files will not be visible in your filesystem (you will see them if you do `help('modules')` though).

`mip` doesn't use `manifest.py`, but rather `package.json`. It could install `.mpy` files if they were listed in the `package.json`.

I don't know how you got a `.mpy` file into your filesystem, though. Perhaps Thonny installed it there or there was some other way?

**Reply by ned-pcs** [reply, marked as answer]
Now that I think about it and have worked on the `mip` code, I think I can answer your question.

Python files inside `micropython-lib` packages get compiled to `.mpy` files as part of the GitHub CI workflow, then get stored to the `gh-pages` branch of `micropython-lib`.

See [this workflow](https://github.com/micropython/micropython-lib/blob/master/.github/workflows/build_packages.yml) and the [CI script](https://github.com/micropython/micropython-lib/blob/master/tools/ci.sh).

The actual `.mpy` files get uploaded to https://micropython.org/pi/v2/file (or your own GitHub module repository if you have opted-in and enabled workflows); see [build.py](https://github.com/micropython/micropython-lib/blob/master/tools/build.py) for a description of the repository layout (there are provisions for the different flavors of `.mpy` files, etc.).

See https://github.com/micropython/micropython-lib/blob/master/CONTRIBUTING.md#publishing-packages-from-forks for a description of how to enable your GitHub module repository to do this compilation automatically upon pushes.
